Make location an optional flag when creating codespaces

jshorty opened the pull request:
- Rely on server-side location determination by default. Both the [machines endpoint](https://docs.github.com/en/rest/reference/codespaces#list-available-machine-types-for-a-repository) and the [creation endpoint](https://docs.github.com/en/rest/reference/codespaces#create-a-codespace-for-the-authenticated-user) itself treat the `location` parameter as optional.
- Allow specifying a location directly with the `l` flag, for instance: `gh codespace create -l WestEurope`
- Remove unused functions (including the API call recommending a location) that has the CLI determine the location up front

I manually tested the following:
- `gh codespace` behaved normally and gave me a `WestUs2` codespace
- `gh codespace create -l WestEurope` behaved normally and gave me a `WestEurope` codespace

jshorty committed:
Location is an optional flag when creating codespaces

jshorty committed:
Drop unused locationResult

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

jshorty commented:
References internal issue codespaces#6924

mislav commented on pkg/cmd/codespace/create.go:
```diff
@@ -38,6 +39,7 @@ func newCreateCmd(app *App) *cobra.Command {
 
 	createCmd.Flags().StringVarP(&opts.repo, "repo", "r", "", "repository name with owner: user/repo")
 	createCmd.Flags().StringVarP(&opts.branch, "branch", "b", "", "repository branch")
+	createCmd.Flags().StringVarP(&opts.location, "location", "l", "", "location (assigned by IP or user settings if not provided)")
```
If we expect non-devs to want to use this flag, would we explain the format of values that this takes?

jshorty replied:
Updated to enumerate the values we'd accept today in the description of this flag— as @josebalius called out we may need to come back and update this later though.

josebalius reviewed (approved):
Nice to see another call get removed! 

Agree with @mislav in that perhaps we should document the valid values for locations. It'll mean that as we add new locations, we'll need to update the docs but I don't expect this happen with much frequency.

jshorty committed:
Enumerate available codespace locations for now

We'll probably want to have an intermediate geographical abstraction for these eventually, but this gives parity with the github.com UI.

mislav commented on pkg/cmd/codespace/create.go:
```diff
@@ -38,6 +39,7 @@ func newCreateCmd(app *App) *cobra.Command {
 
 	createCmd.Flags().StringVarP(&opts.repo, "repo", "r", "", "repository name with owner: user/repo")
 	createCmd.Flags().StringVarP(&opts.branch, "branch", "b", "", "repository branch")
+	createCmd.Flags().StringVarP(&opts.location, "location", "l", "", "location (EastUs, SouthEastAsia, WestEurope, or WestUs2). Assigned by IP or user settings if not provided")
```
We could use the syntax CLI uses for mutually exclusive values:
```suggestion
	createCmd.Flags().StringVarP(&opts.location, "location", "l", "", "location: {EastUs|SouthEastAsia|WestEurope|WestUs2} (determined automatically if not provided)")
```

mislav reviewed (approved):
Looks good!

One alternative to having to update the docs whenever a new location is added could be to add functionality to gh to be able to query and list available locations (like in the old prompt). For example:
- Running `gh cs create --location` (no value) could print the list of available locations in the error message
- Pressing `gh cs create --location <Tab>` could bring up the list of available locations via shell completions. Example: https://github.com/cli/cli/blob/ee3c4aefcb05353d1e08431e36050152b10eab74/pkg/cmd/repo/create/create.go#L178-L196

All this would be extra work, so I'm not sure if anyone feels strongly about having this in.

jshorty committed:
Use proper syntax for location values

Co-authored-by: Mislav Marohnić <mislav@github.com>

jshorty commented:
> One alternative to having to update the docs whenever a new location is added could be to add functionality to gh to be able to query and list available locations (like in the old prompt)

Thanks @mislav — I'm a bit hesitant to hook into the existing locations API for now, since I believe there's a broader internal effort to remove our dependency on it, but will re-evaluate if we find ourselves updating this list much!

Camille0101 commented:
gh pr checkout 5410
